### PR TITLE
Land preview2-prototyping's Wasi Preview 2 implementation into wasmtime

### PR DESCRIPTION
**Pull request by pchickey**
Third task in #6370 

This PR lands preview2-prototyping's Wasi Preview 2 implementation inside `wasmtime_wasi::preview2`.

```
python3 ~/Downloads/git-filter-repo.py --path host --path wasi-common --path test-programs
```
on https://github.com/bytecodealliance/preview2-prototyping/commit/5be99d645c177198b9f5e4d76561669796f7b379

And then merging wasmtime's `pch/test_programs`@87fedb59d6dcf6081285bf390ccf8b5194eaff2c in with --allow-unrelated-histories.

### CODE REVIEW

File: crates/wasi/witx/typenames.witx
Comment (alexcrichton) [inline crates/wasi/witx/typenames.witx:1]: oh man blast from the past

**Review by jameysharp — APPROVED** [review]
I've glanced over `crates/test-programs` and `crates/wasi/src/preview2` but I assume the idea is we want to merge these as work-in-progress and then iterate on them in-tree, right? I see plenty of fixme-type comments and the like. I was particularly curious about the comment in `crates/test-programs/command-tests/src/bin/file_dir_sync.rs` about not being able to open directories as files on Windows.

In particular, it looks like the preview2 implementation doesn't interfere at all with using the existing preview1 implementation. If I've missed any caveats to that I'd like to hear about them, but assuming it's true then I think this is fine to merge, since we aren't making any promises about stability of the component model or preview2 yet.

I've reviewed the rest (`supply-chain`, `ci/run-tests.sh`, and various `Cargo.toml` changes) reasonably carefully.

**Comment by pchickey** [comment]
That's correct: The new implementation is totally separate from the old one. There are a number of known deficiencies with the new implementation but we decided its time to move to this and iterate here. The new implementation won't replace the old one until we have resolved all of those problems, and it passes all of the same tests.

The comment in `file_dir_sync.rs` is a @sunfishcode question - I don't actually know what that means.

**Comment by sunfishcode** [comment]
I think that comment in file_dir_sync.rs is outdated. I think we can uncomment those two lines.

Windows doesn't have a way to do a sync on a directory handle. Originally, WASI's impl would fail on Windows because we'd try to call the Windows API for files and it would fail when passed a directory handle. Now, we handle this so that it silently does nothing, which is the direction I think WASI as a whole is heading. The semantics for `sync` are "do whatever `sync` means in the underlying fs", and on Windows, that means do nothing.

**Pull request merged**